File: leetcode/201-300/288-unique-word-abbreviation/bench/uniqueabbr.py
"""Benchmark twin for LeetCode #288 — same algorithm as uniqueabbr.kara."""

CONFLICTED = object()
LETTERS = "abcdefghijklmnopqrstuvwxyz"


def main() -> None:
    dict_n, pool_n, punches = 3_000, 20_000, 1_000_000
    seed = 12345

    def nxt(s):
        return (s * 1103515245 + 12345) & 2147483647

    words = []
    for _ in range(dict_n):
        seed = nxt(seed)
        n = 3 + ((seed // 65536) % 8)
        cs = []
        for _ in range(n):
            seed = nxt(seed)
            cs.append(LETTERS[(seed // 65536) % 26])
        words.append("".join(cs))

    idx = {}
    for w in words:
        a = w if len(w) <= 2 else f"{w[0]}{len(w) - 2}{w[-1]}"
        prev = idx.get(a)
        if prev is None:
            idx[a] = w
        elif prev is not CONFLICTED and prev != w:
            idx[a] = CONFLICTED

    pool = []
    for i in range(pool_n):
        if i % 2 == 0:
            pool.append(words[(i * 7) % dict_n])
        else:
            seed = nxt(seed)
            n = 3 + ((seed // 65536) % 8)
            cs = []
            for _ in range(n):
                seed = nxt(seed)
                cs.append(LETTERS[(seed // 65536) % 26])
            pool.append("".join(cs))

    unique_count = 0
    for i in range(punches):
        word = pool[i % pool_n]
        a = word if len(word) <= 2 else f"{word[0]}{len(word) - 2}{word[-1]}"
        hit = idx.get(a)
        if hit is None or (hit is not CONFLICTED and hit == word):
            unique_count += 1
    print(f"unique {unique_count}")


if __name__ == "__main__":
    main()

File: leetcode/201-300/288-unique-word-abbreviation/bench/uniqueabbr.rs
// Benchmark twin for LeetCode #288 — same algorithm as uniqueabbr.kara.
use std::collections::HashMap;

// Sole(word) | Conflicted, mirroring the Kara enum.
enum Bucket {
    Sole(String),
    Conflicted,
}

fn abbrev(w: &str) -> String {
    let n = w.len();
    if n <= 2 {
        return w.to_string();
    }
    let b = w.as_bytes();
    format!("{}{}{}", b[0] as char, n - 2, b[n - 1] as char)
}

fn next_rand(state: i64) -> i64 {
    (state * 1103515245 + 12345) & 2147483647
}

fn main() {
    const LETTERS: &[u8] = b"abcdefghijklmnopqrstuvwxyz";
    let dict_n: i64 = 3000;
    let pool_n: i64 = 20000;
    let punches: i64 = 1000000;
    let mut seed: i64 = 12345;

    let mut dict: Vec<String> = Vec::new();
    for _ in 0..dict_n {
        seed = next_rand(seed);
        let n = 3 + ((seed / 65536) % 8);
        let mut w = String::new();
        for _ in 0..n {
            seed = next_rand(seed);
            w.push(LETTERS[((seed / 65536) % 26) as usize] as char);
        }
        dict.push(w);
    }

    let mut idx: HashMap<String, Bucket> = HashMap::new();
    for w in &dict {
        let a = abbrev(w);
        match idx.get(&a) {
            None => {
                idx.insert(a, Bucket::Sole(w.clone()));
            }
            Some(Bucket::Sole(prev)) => {
                if prev != w {
                    idx.insert(a, Bucket::Conflicted);
                }
            }
            Some(Bucket::Conflicted) => {}
        }
    }

    let mut pool: Vec<String> = Vec::new();
    for i in 0..pool_n {
        if i % 2 == 0 {
            pool.push(dict[((i * 7) % dict_n) as usize].clone());
        } else {
            seed = next_rand(seed);
            let n = 3 + ((seed / 65536) % 8);
            let mut w = String::new();
            for _ in 0..n {
                seed = next_rand(seed);
                w.push(LETTERS[((seed / 65536) % 26) as usize] as char);
            }
            pool.push(w);
        }
    }

    let mut unique_count: i64 = 0;
    for i in 0..punches {
        let word = &pool[(i % pool_n) as usize];
        let a = abbrev(word);
        let u = match idx.get(&a) {
            None => true,
            Some(Bucket::Sole(w)) => w == word,
            Some(Bucket::Conflicted) => false,
        };
        if u {
            unique_count += 1;
        }
    }
    println!("unique {}", unique_count);
}
